File: python/block_credentials.py
"""
Credential Theft Blocking Module
Blocks attempts to access environment variables containing secrets
"""
import os
import sys

# Blocked environment variable patterns
BLOCKED_PATTERNS = [
    'KEY', 'SECRET', 'TOKEN', 'PASSWORD', 'CREDENTIAL',
    'API_KEY', 'AUTH', 'PRIVATE', 'ACCESS_TOKEN',
    'GITHUB_TOKEN', 'OPENAI_API_KEY', 'ANTHROPIC_API_KEY'
]

class SecureEnviron:
    """Proxy that blocks access to sensitive env vars"""
    def __getitem__(self, key):
        key_upper = key.upper()
        for pattern in BLOCKED_PATTERNS:
            if pattern in key_upper:
                raise PermissionError(f"Blocked: Attempt to access credential {key}")
        return os.environ[key]
    
    def get(self, key, default=None):
        try:
            return self[key]
        except (KeyError, PermissionError):
            return default

# Patch os.environ
os.environ = SecureEnviron()

# Block common credential theft vectors
BLOCKED_PATHS = [
    '/root/.aws',
    '/root/.ssh',
    '/root/.config',
    '/home/*/.aws',
    '/home/*/.ssh', 
    '/home/*/.config',
    '/etc/passwd',
    '/etc/shadow',
    '/etc/security/opasswd'
]

def block_file_access(path):
    """Block access to sensitive files"""
    import builtins
    original_open = builtins.open
    
    def restricted_open(path, *args, **kwargs):
        for blocked in BLOCKED_PATHS:
            if path.startswith(blocked.replace('*', '')):
                raise PermissionError(f"Blocked: Access to {path} is not allowed")
        return original_open(path, *args, **kwargs)
    
    builtins.open = restricted_open

# Apply restrictions
try:
    block_file_access(None)
except:
    pass

print("[ASF Security] Credential theft protection enabled")

File: nodejs/block_credentials.js
/**
 * ASF Credential Theft Blocking Module for Node.js
 * Blocks attempts to access environment variables containing secrets
 */

const originalEnv = process.env;
const blockedPatterns = [
  'KEY', 'SECRET', 'TOKEN', 'PASSWORD', 'CREDENTIAL',
  'API_KEY', 'AUTH', 'PRIVATE', 'ACCESS_TOKEN',
  'GITHUB_TOKEN', 'OPENAI_API_KEY', 'ANTHROPIC_API_KEY'
];

// Blocked file paths
const blockedPaths = [
  '/root/.aws',
  '/root/.ssh',
  '/root/.config',
  '/home/',
  '/etc/passwd',
  '/etc/shadow',
  '/etc/security/'
];

// Block sensitive environment variables
Object.defineProperty(process, 'env', {
  get: () => {
    const env = {};
    for (const key in originalEnv) {
      const upperKey = key.toUpperCase();
      let blocked = false;
      for (const pattern of blockedPatterns) {
        if (upperKey.includes(pattern)) {
          blocked = true;
          break;
        }
      }
      if (!blocked) {
        env[key] = originalEnv[key];
      }
    }
    return env;
  },
  configurable: false
});

// Block file system access to sensitive paths
const originalReadFile = require('fs').readFile;
const originalReadFileSync = require('fs').readFileSync;
const fs = require('fs');

function checkPath(path) {
  for (const blocked of blockedPaths) {
    if (path.startsWith(blocked)) {
      throw new Error(`ASF Security: Blocked access to ${path}`);
    }
  }
}

const secureFs = new Proxy(fs, {
  get(target, prop) {
    if (['readFile', 'readFileSync', 'open', 'openSync', 'readdir', 'readdirSync'].includes(prop)) {
      return function(...args) {
        if (args[0]) checkPath(args[0]);
        return target[prop](...args);
      };
    }
    return target[prop];
  }
});

// Replace fs module for skill execution
Object.assign(global, { fs: secureFs });

console.log('[ASF Security] Node.js credential theft protection enabled');
